design_docs/selfhealing_featuretoggle.md

```diff
@@ -0,0 +1,160 @@
+
+```
+keptn create project PROJECTNAME --shipyard=shipyard.yaml
+```
+
+
+
+
+
+### Onboard a service
+
+Onboard the service and add the remediation file as a resource.
+
+
+```
+keptn onboard service SERVICENAME --project=PROJECTNAME --chart=HELMCHART
+```
+
+
+Configure Keptn - add the remediation file
+
+
+```
+keptn add-resource --project=PROJECTNAME --service=SERVICENAME --stage=STAGENAME --resource=remediation.yaml
+```
+
+Deploy your application to Keptn.
+
+```
+keptn send event new-artifact --project=PROJECTNAME --service=SERVICENAME --image=docker.io/IMAGENAME --tag=TAG
+```
+
+</details>
+
+
+### Add secret for Unleash server
+
+In order to let the Keptn built-in Unleash remediation service communicate with the Unleash server, provide the credentials in terms of a Kubernetes secret.
+
+unleash-secret.yaml:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: unleash-secret
+type: Opaque
+data:
+  username: unleash-username
+  token: unleash-token
+  server-url: unleash-server-url
+```
+
+
+Apply the secret to the cluster.
+
+```
+kubectl apply -f unleash-secret.yaml -n keptn
+```
+
+<!--
+ALTERNATIVE
+
+```
+#keptn add secret secret.yaml # needed to access the unleash server
+#keptn apply uniform --project=PROJECTNAME uniform.yaml
+```
+
+uniform not needed right now since unleash is built in as action in Keptn core remediation service 
+-->
+
+### Configure your monitoring
+
+To be able to react on any issues in your environment, the monitoring solution has to be setup and configured to send details of the detected issue to Keptn. In case of Dynatrace, this is how it would look like:
+
+1. Follow the Keptn docs to setup Dynatrace [as described here](https://keptn.sh/docs/0.5.0/reference/monitoring/dynatrace/).
+
+2. Set up a custom alert rule to be notified if the 90th percentile of the response time of your service is slower than expected. This custom alert rule does very much depend on your [SLOs](https://github.com/keptn/spec/blob/master/sre.md). 
+
+    An example can be seen in this screenshot:
+    ![Custom alert](./assets/dt-custom-alert.png)
+
+    Please note that in this case, the title of the custom alert has to match the remediation name in the remediation.yaml file, i.e., in this case `responsetime_p90_sockshop_carts`.
+
+
+
+### Watch self healing in action
+
+Now, if the response time of your service is slower than defined in your SLOs, the monitoring tool will send out an alert (problem ticket) to Keptn. Keptn on the other hand will find that there is a remediation set for the affected service (defined in the remediation.yaml file). Thus, Keptn will trigger the remediation, by making an API call to the feature toggle server and set the toggle to the desired state described in the remediation action. Afterwards, Keptn will send out a `sh.keptn.events.deployment-finished` event (TODO verify what is correct event type) to verify via the Keptn quality gates if the remediation access was successful. 
```

Comment:
Let's discuss the correct events in one of the next spec meetings and also how to handle the verification of remediation actions.